### PR TITLE
Error early when dataclass is not registered

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112211

Partially fixes: https://github.com/pytorch/pytorch/issues/112043
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @angelayi @suo

### CODE REVIEW

File: test/dynamo/test_export.py [hunk 1]
```diff
@@ -4238,6 +4239,25 @@ def forward(self, x):
...
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, input: Input):
+                return input.foo + input.foo
+
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "Perhaps you forgot to register it as dataclass. "
+            "Please take a look at torch._export.utils.register_dataclass_as_pytree_node.",
```
Comment (zhxchen17) [inline test/dynamo/test_export.py:4257]:
```suggestion
            "Dataclasses are supposed to be pytree nodes to be exportable."
            "Please take a look at torch._export.utils.register_dataclass_as_pytree_node for more information.",
```

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -1201,6 +1202,22 @@ def result_capturing_wrapper(*graph_inputs):
         # optimize_assert since user program may mutate the inputs.
         flat_args, in_spec = pytree.tree_flatten((args, kwargs))
 
+        for arg in flat_args:
+            prim_types = (torch.Tensor, str, int, float, bool)
+            if not isinstance(arg, prim_types):
+                if dataclasses.is_dataclass(arg):
+                    raise RuntimeError(
+                        "Dataclasses are supposed to be pytree nodes to be exportable."
+                        "Please take a look at torch._export.utils.register_dataclass_as_pytree_node "
```
Comment (suo) [inline torch/_dynamo/eval_frame.py:1211]: depressing that we're suggesting private APIs to people

Comment (zhxchen17) [reply torch/_dynamo/eval_frame.py:1211]: yeah actually we should just publish this API. @tugsbayasgalan

Comment (angelayi) [reply torch/_dynamo/eval_frame.py:1211]: It is public [`torch.export.register_dataclass`](https://fburl.com/code/hg2la1lc)

**Review by angelayi — APPROVED** [review]

**Comment by ZainRizvi** [comment]
@tugsbayasgalan PTAL

**Pull request closed without merging**